src/box/sql/show.c

```diff
@@ -0,0 +1,494 @@
+	sql_desc_append(desc, ")");
+}
+
+/** Add a tuple foreign key constraint to the statement description. */
+static void
+sql_describe_tuple_foreign_key(struct sql_desc *desc,
+			       const struct space_def *def,
+			       const struct tuple_constraint_def *cdef)
+{
+	if (desc->is_aborted)
+		return;
+	assert(cdef->type == CONSTR_FKEY && cdef->fkey.field_mapping_size > 0);
+	const struct space *foreign_space = space_by_id(cdef->fkey.space_id);
+	assert(foreign_space != NULL);
+	bool is_error = false;
+	for (uint32_t i = 0; i < cdef->fkey.field_mapping_size; ++i) {
+		const struct tuple_constraint_field_id *field =
+			&cdef->fkey.field_mapping[i].local_field;
+		if (field->name_len == 0 &&
+		    field->id >= foreign_space->def->field_count) {
+			sql_desc_error(desc, "foreign key", cdef->name,
+				       "local field is unnamed");
+			is_error = true;
+		}
+		field = &cdef->fkey.field_mapping[i].foreign_field;
+		if (field->name_len == 0 &&
+		    field->id >= foreign_space->def->field_count) {
+			sql_desc_error(desc, "foreign key", cdef->name,
+				       "foreign field is unnamed");
+			is_error = true;
+		}
+	}
+	if (is_error)
+		return;
+
+	sql_desc_append(desc, ",\nCONSTRAINT ");
+	sql_desc_append_name(desc, cdef->name);
+	sql_desc_append(desc, " FOREIGN KEY(");
+	for (uint32_t i = 0; i < cdef->fkey.field_mapping_size; ++i) {
+		const struct tuple_constraint_field_id *field =
+			&cdef->fkey.field_mapping[i].local_field;
+		const char *field_name = field->name_len != 0 ? field->name :
+					 def->fields[field->id].name;
+		if (i > 0)
+			sql_desc_append(desc, ", ");
+		sql_desc_append_name(desc, field_name);
+	}
+
+	sql_desc_append(desc, ") REFERENCES ");
+	sql_desc_append_name(desc, foreign_space->def->name);
+	sql_desc_append(desc, "(");
+	for (uint32_t i = 0; i < cdef->fkey.field_mapping_size; ++i) {
+		const struct tuple_constraint_field_id *field =
+			&cdef->fkey.field_mapping[i].foreign_field;
+		assert(field->name_len != 0 || field->id < def->field_count);
+		const char *field_name = field->name_len != 0 ? field->name :
+					 def->fields[field->id].name;
+		if (i > 0)
+			sql_desc_append(desc, ", ");
+		sql_desc_append_name(desc, field_name);
+	}
+	sql_desc_append(desc, ")");
+}
+
+/** Add a field check constraint to the statement description. */
+static void
+sql_describe_field_check(struct sql_desc *desc, const char *field_name,
+			 const struct tuple_constraint_def *cdef)
+{
+	if (desc->is_aborted)
+		return;
+	assert(cdef->type == CONSTR_FUNC);
+	bool is_error = false;
+	const struct func *func = func_by_id(cdef->func.id);
+	if (func->def->language != FUNC_LANGUAGE_SQL_EXPR) {
+		sql_desc_error(desc, "check constraint", cdef->name,
+			       "wrong constraint expression");
+		is_error = true;
+	} else if (!func_sql_expr_has_single_arg(func, field_name)) {
+		sql_desc_error(desc, "check constraint", cdef->name,
+			       "wrong field name in constraint expression");
+		is_error = true;
```

Comment:
No, this is actually different. In ANSI query
```
CREATE TABLE t(i INT PRIMARY KEY, a INT CHECK(i > 10));
```
is the same as
```
CREATE TABLE t(i INT PRIMARY KEY, a INT, CHECK(i > 10));
```
since all column constraints are just another syntax for table constraints. This is not so for us. For example, we can set constraint "i > 10" in Lua for field `a` and it will check field `a`, since no names are checked in core field constraints. However, if we allow `SHOW CREATE TABLE` for such cases we receive
```
CREATE TABLE t(i INT PRIMARY KEY, a INT CHECK(i > 10));
```
which could not be executed from out standpoint and will have another result for ANSI standpoint.

If only we were able to avoid so-called "field constraints", we wouldn't have such problems. But we seek for speed here, it seems.

Reply:
So like I said, it can't be executed anyway. But you still care to validate it. Which means this is exactly like the situation with fieldno vs name was, and exactly like the tuple constraints which currently are not validated and won't be executed either (I gave you the example above). I don't see why keep this particular validation, if others are not present or don't help anyway, just like this one.

Reply:
I thought we decided that if the result was returned by `SHOW CREATE TABLE tablename;`, then that means the result should be executable. If I skip the check, this rule may be violated. Not sure if this is correct.

Reply:
This is already violated. I gave you an example above. Can copy-paste it again:
```Lua
box.cfg{}
body = 'A > 100 AND A < 200'
box.schema.func.create('check', {is_deterministic = true, body = body, language = 'SQL_EXPR'})
s = box.schema.create_space('test', {
    format = {{'ID', 'unsigned'}},
    constraint = {check = 'check'}})
_ = s:create_index('pk')
box.execute('SHOW CREATE TABLE "test"')
```
```
---
- metadata:
  - name: statement
    type: string
  rows:
  - ['CREATE TABLE "test"(

      ID UNSIGNED NOT NULL,

      CONSTRAINT "pk" PRIMARY KEY(ID),

      CONSTRAINT "check" CHECK(A > 100 AND A < 200));']
...
```
The returned SQL statement can't be executed. It will throw an error. That means the `SHOW CREATE TABLE` returned a bad result.

Reply:
I will fill a bug. I am not sure which of these are more correct.